### PR TITLE
Allow `(get-model)` after answering `unknown`

### LINKED ISSUE #8854: No more models when resource limit is hit
As part of our migration from CVC4 to cvc5 in both Why3 and SPARK, we've hit a problem that seems to come from a change in how cvc5 handles models when resource limit (time or steps) is hit. CVC4 used to display a model, which we could compare with other models (calling the prover with more or less context), as well as filter valid models by executing them with different semantics (as described in the article "Explaining Counterexamples with Giant-Step Assertion Checking" at F-IDE 2021).

Instead, cvc5 displays the following message instead of a model when a resource limit is hit:

```
Cannot get model since model is not available. Perhaps the most recent call to check-sat was interrupted?
```

Is there a switch to restore the previous behavior?
Otherwise, could such a functionality be provided?
Otherwise, we won't be able to use cvc5 for getting models ("counterexamples" in our case) when proof fails.

Ideally of course, we'd prefer cvc5 to validate as much as possible the partial model obtained when a resource limit is hit, against ground facts, before displaying it, so that we don't have to filter bad models out. But discarding all partial models is not a suitable solution in our case.

### PR DESCRIPTION
**Pull request by 4tXJ7f**
Fixes #8854. Commit a407bd70c09724dc20af3241775abedd3a73ec67
accidentally removed our ability to print models after answering
`unknown`. Before that commit, we were able to print the models when we
had a theory that required a last effort check. In that case, we were
generating models periodically that would then get printed. This commit
restores this ability, simplifies `TheoryEngine`, and attempts to build
a model if none exists. I have tested multiple theories and examples and
it seems to work fairly well.

### CODE REVIEW

File: src/theory/theory_engine.h [hunk 1]
```diff
@@ -349,20 +349,6 @@ class TheoryEngine : protected EnvObj
...
-   * (preprocessed) input formula and all theory lemmas that have been marked
-   * NEEDS_JUSTIFY. For more details on this, see relevance_manager.h.
-   *
-   * This method updates success to false if the set of relevant assertions
-   * is not available. This may occur if we are not in SAT mode, if the
-   * relevance manager is disabled (see option::relevanceFilter) or if the
-   * relevance manager failed to compute relevant assertions due to an internal
-   * error.
-   */
-  std::unordered_set<TNode> getRelevantAssertions(bool& success);
```
Comment (ajreynol) [inline src/theory/theory_engine.h:364]: Can you keep this? I had a request for this functionality at some point but haven't connected it to the API yet.

**Review by ajreynol — CHANGES_REQUESTED** [review]
Can you update the documentation in `theory_engine.h` for `getBuiltModel`? I think it allows returning the model even when interrupted now?

File: src/theory/theory_engine.h (hunk 1 above)
Comment (4tXJ7f) [reply src/theory/theory_engine.h]: Sounds good, I've added it back. I've changed it slightly (to not depend on `d_inSatMode` and moved it to the same relative place as its declaration in the header file.

**Comment by ajreynol** [comment]
It wasn't clear to me what exactly the idea of this PR was at first.  However, I think I see: the key is that when we are interrupted, the model is that last one generated, which is likely even from a *different SAT context*.  Is that right? 

If so, I think this is a really nice.  This is way simpler than having the interrupt wait until we generate the *next* model.  

Can you update the documentation to make this clear? A key thing to mention is that the model is SAT context-independent internally.

I have 2 concerns:
(1) what happens when no model has been built?
(2) what if we are interrupted *while* the model in being constructed? 

I think (1) should be harmless, we return the default model. For (2), I suppose there are ways to safeguard this?

**Pull request merged**